### PR TITLE
ENH: control PeakStats in spec API via gs / gs.SUB_FACTORIES refactor

### PR DESCRIPTION
**Pull request by tacaswell**
Closes https://github.com/NSLS-II/Bug-Reports/issues/123

attn @ambarb 

You can control this via `gs.PS_CONFIG['edge_count'] = n`, 0 turns it off.

### CODE REVIEW

File: bluesky/spec_api.py [hunk 1]
```diff
@@ -29,8 +29,10 @@
 import itertools
 from itertools import chain
 from collections import ChainMap
+from inspect import signature, Parameter
+import warnings
 
-### Factory functions for generating callbacks
+# ## Factory functions for generating callbacks
```
Comment (danielballan) [inline bluesky/spec_api.py]: Ha.

File: bluesky/spec_api.py [hunk 2]
```diff
@@ -54,7 +56,7 @@ def _figure_name(base_name):
     return base_name
 
 
-def setup_plot(motors):
+def setup_plot(motors, dets, gs):
```
Comment (danielballan) [inline bluesky/spec_api.py]: Note in docstring that dets, gs are unused args.

File: bluesky/spec_api.py [hunk 3]
```diff
@@ -71,15 +73,61 @@ def setup_plot(motors):
         return LivePlot(y_key, fig=fig)
 
 
-def setup_peakstats(motors):
+def setup_peakstats(motors, dets, gs):
```
Comment (danielballan) [inline bluesky/spec_api.py]: Note in docstring that `dets` is unused.

File: bluesky/spec_api.py [hunk 4]
```diff
@@ -71,15 +73,61 @@ def setup_plot(motors):
...
+    motor = list(motors)[0]
+    key = first_key_heuristic(motor)
+    ps = PeakStats(key, gs.MASTER_DET_FIELD, **gs.PS_CONFIG)
     gs.PS = ps
+    ps.motor = motor
     return ps
 
 
-### Counts (p. 140) ###
+def setup_livetable(motors, dets, gs):
```
Comment (danielballan) [inline bluesky/spec_api.py]: ditto

File: doc/source/simple_api.rst [hunk 5]
```diff
@@ -97,10 +97,18 @@ There are other settings which control the output of the scans --
 ``gs.TABLE_COLS`` and ``gs.PLOT_Y``  for example. Explore the contents of
 ``gs`` by typing ``gs.<TAB>``.
 
+Peak Stats
+----------
```
Comment (danielballan) [inline doc/source/simple_api.rst]: A+ docs.

**Comment by danielballan** [comment]
I really like this refactor of the SPEC subscriptions. See some minor notes and add content to docs.

**Comment by tacaswell** [comment]
I do not think the docstrings need notes on if they internals are used as they are conforming to a specified API.

File: bluesky/spec_api.py [hunk 6]
```diff
@@ -71,15 +73,61 @@ def setup_plot(motors):
...
+def setup_liveraster(motors, dets, gs, *, shape, extent):
+    if len(motors) != 2:
+        return None
+    ylab, xlab = [first_key_heuristic(m) for m in motors]
+    raster = LiveRaster(shape, gs.MASTER_DET_FIELD, xlabel=xlab,
+                        ylabel=ylab, extent=extent, clim=[0, 1])
+    return raster
+
+
+def _construct_subs(plan_name, motors, dets, **kwargs):
```
Comment (klauer) [inline bluesky/spec_api.py]: I think this should be public, since I should be able to write other `gs`-abiding scans if I so choose.

**Comment by klauer** [comment]
👍  on the functionality. 

Any thoughts on changing the `setup_*` signature to required kwargs instead of positional args?

File: bluesky/spec_api.py (hunk 1 above)
Comment (klauer) [inline bluesky/spec_api.py]: Every time you laugh at anything PEP8-related, you are required to toss a quarter in the jar in @hhslepicka's office.

**Comment by tacaswell** [comment]
:+1: to required kwargs!  Lets see how far we can push that before it gets too annoying :smiling_imp:

**Comment by tacaswell** [comment]
these commits are squashed and moved into #516

**Pull request closed without merging**